compiler/qsc_partial_eval/tests/operators.rs

```diff
@@ -300,3 +303,2065 @@ fn logical_not_unary_operator_generates_logical_not_rir_instruction() {
+                Variable(2, Boolean) = Icmp Eq, Variable(0, Boolean), Variable(1, Boolean)
+                Call id(3), args( Variable(2, Boolean), Pointer, )
+                Return"#]],
+    );
+    assert_eq!(program.num_qubits, 2);
+    assert_eq!(program.num_results, 2);
+}
+
+#[test]
+fn comparing_measurement_results_for_inequality_adds_read_result_and_comparison_instructions() {
+    let program = get_rir_program(indoc! {
+        r#"
+        namespace Test {
+            @EntryPoint()
+            operation Main() : Bool {
+                use (q0, q1) = (Qubit(), Qubit());
+                let r0 = MResetZ(q0);
+                let r1 = MResetZ(q1);
+                r0 != r1
+            }
+        }
+        "#,
+    });
+    let measurement_callable_id = CallableId(1);
+    assert_callable(
+        &program,
+        measurement_callable_id,
+        &expect![[r#"
+        Callable:
+            name: __quantum__qis__mresetz__body
+            call_type: Measurement
+            input_type:
+                [0]: Qubit
+                [1]: Result
+            output_type: <VOID>
+            body: <NONE>"#]],
+    );
+    let readout_callable_id = CallableId(2);
+    assert_callable(
+        &program,
+        readout_callable_id,
+        &expect![[r#"
+        Callable:
+            name: __quantum__qis__read_result__body
+            call_type: Readout
+            input_type:
+                [0]: Result
+            output_type: Boolean
+            body: <NONE>"#]],
+    );
+    let bool_record_id = CallableId(3);
+    assert_callable(
+        &program,
+        bool_record_id,
+        &expect![[r#"
+        Callable:
+            name: __quantum__rt__bool_record_output
+            call_type: OutputRecording
+            input_type:
+                [0]: Boolean
+                [1]: Pointer
+            output_type: <VOID>
+            body: <NONE>"#]],
+    );
+    assert_block_instructions(
+        &program,
+        BlockId(0),
+        &expect![[r#"
+            Block:
+                Call id(1), args( Qubit(0), Result(0), )
+                Call id(1), args( Qubit(1), Result(1), )
+                Variable(0, Boolean) = Call id(2), args( Result(0), )
+                Variable(1, Boolean) = Call id(2), args( Result(1), )
+                Variable(2, Boolean) = Icmp Ne, Variable(0, Boolean), Variable(1, Boolean)
+                Call id(3), args( Variable(2, Boolean), Pointer, )
+                Return"#]],
+    );
+    assert_eq!(program.num_qubits, 2);
+    assert_eq!(program.num_results, 2);
+}
+
+#[test]
+fn comparing_measurement_result_against_result_literal_for_equality_adds_read_result_and_comparison_instructions(
+) {
+    let program = get_rir_program(indoc! {
+        r#"
+        namespace Test {
+            @EntryPoint()
+            operation Main() : Bool {
+                use q = Qubit();
+                let r = MResetZ(q);
+                r == One
```

Comment:
With the suggestion about optimizing some comparisons, this case will become a store, which unfortunately means we'd want a `r == Zero` case to confirm that the right comparison operation emits for that. We'd also need an `r != One` case to complement the one below this.

Reply:
Updated the tests and added tests for when both operands are dynamic.